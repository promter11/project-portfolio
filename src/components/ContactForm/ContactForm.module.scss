@import "../../styles/_variables";
@import "../../styles/_mixins";

.form {
  @include flex(wrap, row, space-between, flex-start);

  .field {
    width: calc(50% - 1.5rem);
    margin-bottom: 3rem;

    @media (max-width: 1200px) {
      width: 100%;
    }

    &.fullWidth {
      width: 100%;
    }

    .text {
      display: block;
      margin-bottom: 1rem;
      @include text(1.4rem, bold, 1, uppercase);
      color: $black-color;
    }

    .label {
      .input,
      .textarea {
        width: 100%;
        transition: border 0.25s ease;

        &:hover {
          border-bottom: 0.1rem solid $contact-block-color;
        }

        &:focus {
          border: 0.1rem solid $contact-block-color;
        }
      }

      .input,
      .textarea {
        font-family: "Museo Sans", sans-serif;
        @include text(1.4rem, 300, 150%, none);
        padding: 1rem 1.5rem;
        border: 0.1rem solid $form-border-color;
        outline: none;
        background-color: transparent;
        color: $black-color;
      }

      .textarea {
        height: 12rem;
        resize: none;
      }
    }

    .error {
      display: inline-block;
      margin-top: 0.5rem;
      color: $error-color;
    }
  }

  .submit {
    @include button($white-color, $black-color);
  }
}
