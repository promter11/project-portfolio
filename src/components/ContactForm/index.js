import React from "react";
import { Form, Field } from "react-final-form";

import { contactText } from "../../common/jsonText/contactText";

import styles from "./ContactForm.module.scss";

const composeValidators = (...validators) => (value) =>
  validators.reduce((error, validator) => error || validator(value), undefined);

const onSubmit = (values) => console.log(JSON.stringify(values));

const ContactForm = ({ language }) => {
  const required = (value) =>
    value ? undefined : contactText[language].blocks.right.form.errors.required;
  const maxLength = (max) => (value) =>
    value.length <= max
      ? undefined
      : contactText[language].blocks.right.form.errors.maxLength;
  const checkEmail = (pattern) => (value) =>
    pattern.test(value)
      ? undefined
      : contactText[language].blocks.right.form.errors.checkEmail;

  return (
    <Form
      onSubmit={onSubmit}
      render={({ handleSubmit, form, submitting }) => (
        <form className={styles.form} onSubmit={handleSubmit}>
          <Field name="name" validate={required}>
            {({ input, meta }) => (
              <div className={styles.field}>
                <span className={styles.text}>
                  {contactText[language].blocks.right.form.name}
                </span>
                <label className={styles.label}>
                  <input {...input} className={styles.input} type="text" />
                </label>
                {meta.error && meta.touched && (
                  <span className={styles.error}>{meta.error}</span>
                )}
              </div>
            )}
          </Field>
          <Field
            name="email"
            validate={composeValidators(
              required,
              maxLength(64),
              checkEmail(/^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$/)
            )}
          >
            {({ input, meta }) => (
              <div className={styles.field}>
                <span className={styles.text}>
                  {contactText[language].blocks.right.form.email}
                </span>
                <label className={styles.label}>
                  <input {...input} className={styles.input} type="text" />
                </label>
                {meta.error && meta.touched && (
                  <span className={styles.error}>{meta.error}</span>
                )}
              </div>
            )}
          </Field>
          <Field
            name="message"
            validate={composeValidators(required, maxLength(255))}
          >
            {({ input, meta }) => (
              <div className={`${styles.field} ${styles.fullWidth}`}>
                <span className={styles.text}>
                  {contactText[language].blocks.right.form.message}
                </span>
                <label className={styles.label}>
                  <textarea {...input} className={styles.textarea} />
                </label>
                {meta.error && meta.touched && (
                  <span className={styles.error}>{meta.error}</span>
                )}
              </div>
            )}
          </Field>
          <button className={styles.submit} type="submit" disabled={submitting}>
            {contactText[language].blocks.right.form.submit}
          </button>
        </form>
      )}
    />
  );
};

export default ContactForm;
