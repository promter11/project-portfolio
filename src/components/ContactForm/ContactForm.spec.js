import React from "react";
import { shallow } from "enzyme";

import ContactForm from "./index";

const props = {
  language: "RU",
};

describe("ContactForm component", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<ContactForm {...props} />);
  });

  it("renders correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe("Functions", () => {
    it("should call setModalStatus function", () => {
      wrapper.find("Modal").props().toggleModal();

      expect(wrapper.find("Modal").prop("modalStatus")).toEqual(true);
    });

    it("should submit form", () => {
      wrapper.find("ReactFinalForm").simulate("submit");

      expect(wrapper.find("Modal").prop("modalStatus")).toEqual(true);
    });
  });
});
