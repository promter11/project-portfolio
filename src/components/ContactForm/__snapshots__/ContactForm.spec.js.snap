// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactForm component renders correctly 1`] = `
<Fragment>
  <ReactFinalForm
    onSubmit={[Function]}
    render={[Function]}
  />
  <Modal
    modalStatus={false}
    toggleModal={[Function]}
  >
    <div
      className="modalInner"
    >
      <Heading
        className="modalTitle"
        level={2}
      >
        Спасибо!
      </Heading>
      <p
        className="modalDesc"
      >
        Ваше сообщение успешно отправлено.
      </p>
    </div>
  </Modal>
</Fragment>
`;
